align_system/utils/incontext_utils.py

```diff
@@ -0,0 +1,287 @@
+                # Load each example in the dataset file
+                for icl_sample in dset:
+                    # Get state and actions
+                    state, actions = hydrate_scenario_state(icl_sample["input"])
+                    # Get choices
+                    choices = adm_utils.format_choices(
+                        [a.unstructured for a in actions],
+                        actions,
+                        state,
+                        self.log
+                    )
+                    # Get KDMA_values
+                    kdma_values = []
+                    for label in icl_sample["label"]:
+                        if sys_kdma_name not in label:
+                            kdma_values.append(None)
+                        else:
+                            kdma_values.append(label[sys_kdma_name])
+                    example = {'state':state, 'actions': actions, 'choices':choices, 'kdma_values':kdma_values}
+                    incontext_data[sys_kdma_name].append(example)
+            
+            # TODO - add KDMA normalization option
+            
+            return incontext_data
+
+    def select_icl_examples(self, sys_kdma_name, prompt_to_match):
+        '''
+        Selects a list of relevant ICL examples
+        Input:
+            sys_kdma_name - key of the target kdma in self.icl_datasets
+            prompt_to_match - the prompt we are selecting ICL examples for
+        Output:
+            selected_icl_examples - relevant subset of self.icl_datasets
+        '''
+        # Check that we have incontext examples for the target kdma
+        if sys_kdma_name not in self.icl_datasets:
+            raise RuntimeError(f"No incontext samples for targeted kdma: {sys_kdma_name}")
+        n_icl_examples = self.incontext_settings["number"]
+        possible_icl_examples = self.icl_datasets[sys_kdma_name]
+        # Check that we have enough incontext examples for the target kdma
+        if len(possible_icl_examples) < n_icl_examples:
+            raise RuntimeError(f"Not enough possible incontext samples to learn from. Only "
+                            f"{len(possible_icl_examples)} samples available while asking for "
+                            f"{n_icl_examples} incontext samples.")
+        # If using LOO, don't include example ICL with exact same prompt
+        if self.incontext_settings.get("leave_one_out", False):
+            possible_icl_examples = [
+                icl_ex for icl_ex in possible_icl_examples
+                if icl_ex["prompt"] != prompt_to_match
+            ]
+
+        # Downselect to n_icl_examples via given method
+        icl_strategy = self.incontext_settings["method"]
+        
+        if icl_strategy == "random":
+            selected_icl_examples = random.sample(possible_icl_examples, n_icl_examples)
+        elif icl_strategy == "bert_similarity":
+            possible_icl_prompts = [icl_sample["prompt"] for icl_sample in possible_icl_examples]
```

Comment:
Similarly, there is a use case where the prompt we use for checking similarity might be slightly different than the prompt we want to present to the ADM in the end. The easy example for this is that maybe we only want to condition similarity on state description, not available actions. I do this in the [base Outlines ADM](https://github.com/ITM-Kitware/align-system/blob/main/align_system/algorithms/outlines_adm.py#L312-L317)

It might be a good check to try to implement that ICL method as well to make sure this interface is flexible enough

Reply:
Good point, I will work on the base outlines ADM ICL implementation. I am thinking we should have two bert similarity strategy options - one for prompt matching and one for scenario description matching. That will require the change I described in the comment above so we can access the scenario descriptions

Reply:
I've added the two bert similarity options here https://github.com/ITM-Kitware/align-system/pull/115/commits/f4fdf9060b9ef0d5498e5ee691f060bdd0469f56
I'll reformat the base outlines ADM ICL next week

Reply:
I've reformatted the base outlines ADM ICL here https://github.com/ITM-Kitware/align-system/pull/115/commits/44cb70cb684cfb51c6bdd51faec3ed631a90dbee

Reply:
See [related comment](https://github.com/ITM-Kitware/align-system/pull/115#discussion_r1742263054)